src: allow for year = 0; BC --> BCE; facilitate Astronomical and (B)CE timeline (#1580)
c3d/db48x: doc/commands/system.md

# Time, Date, Alarms and System Commands

## DateMenu

> Activated from the [Time](##TimeMenu) menu.

Reviewer: Access path is valuable, but should be last. The description is more important.

Reviewer: Thinking more about it, there is a lot of demand for a way to know the access path in the documentation. I filed #1622 for that.

Author: moved it

> # Date, Time

Reviewer: ```suggestion
# Date and Time
```

Author: done

## Date entry / edit format

A whole number with a `_date` unit attached is interpreted as a date according to the customary format `YYYYMMDD`.
However, the year need not be a 4 digit number.
It can a smaller, bigger, zero, even a negative number and leading zeros can be ommitted
(as such it need not fit the `YYYY` format).
In case of a negative number only the year is treated as such.
Months and days, even in a negative year, run from low (positive numbers) to high within the year.

Note: the date format is intentionally different from the format on the HP-48.

Note: not all commands can handle a negative year.

> The underlying platform DMCP of the SwissMicros calculators handles only positive years.

Reviewer: ```suggestion
The DMCP platform for SwissMicros calculators only handles positive years. This restriction bubbles up to several DB48x commands.
```

Author: done

That bubbles up to several commands.

## Date display format

By default a Date is displayed on the stack like `Sun 30/Nov/2025`.
The format can be adjusted by changing the [Date, Time flags](#date-time-flags) or
[Date, Time settings](#date-time-settings).

## Date+Time entry / edit format

> A real number with a `_date` unit attached is interpreted as a Date + Time

Reviewer: ```suggestion
A decimal number with a `_date` unit attached is interpreted as a Date + Time
```

Author: done

according to the format `YYYYMMDD.HHMMSS`.

## Date+Time display format

By default a Date+Time is displayed on the stack like `Sun 30/Nov/2025, 16:43:10`.
The format can be adjusted by changing the [Date, Time flags](#date-time-flags) or
[Date, Time settings](#date-time-settings).

> ## Time entry format

Reviewer: ```suggestion
## Data entry for time or angle values
```

Author: Want to say as little as possible about angles here.
Kept the title.

> Enter the hours, a '.' to end the hours an begin the minutes.

Reviewer: This is already documented under `## Angle operations` in `2-Quickstart.md`.

It is worth repeating here, but I think that the wording can be improved. What about:

DB48x features a method to quickly enter angle or time values with repeated uses of the decimal dot. For example, to enter `2°03′05″`, you can type the key sequence:
| _2_ | _._ | _3_ | _._ | _5_ |

At this point the editor shows `2°3′5_dms`. Hitting _ENTER_ will then put `2°03′05″` on the stack. Additional uses of the decimal dot make it possible to enter fractions of a second.

If you want to enter a time value instead of an angle, you can use the right arrow to put the cursor on the unit, and then cycle between units using the _EEX_ (`Cycle`) key to alternate between `_dms` and `_hms`.

Author: merged your suggestions

## Time, Date related settings

* `DateSlash` / `DateDash` / `DateDot` / `DateSpace`

## BCECEyearNumbering

Activate the B(efore) C(ommon) E(ra) / C(ommon) E(ra) suffix of the year in the display of a date on the stack.
Activate the related year numbering which jumps from 1 BCE to 1 CE. So negative years are shifted.
See: [Proleptic Gregorian calendar](https://en.wikipedia.org/wiki/Proleptic_Gregorian_calendar)

## AstronomicalYearNumbering

Activate the Astronomical year numbering. This uses a uniform timeline including a year = 0.
Display negative years on the stack with a minus sign.
See: [Astronomical year numbering](https://en.wikipedia.org/wiki/Astronomical_year_numbering)

## Date entry / edit format

A whole number with a `_date` unit attached is interpreted as a date according to the customary format `YYYYMMDD`.
The year need not be a 4 digit number.
It can a smaller, bigger, zero, even a negative number and leading zeros can be ommitted
(as such it need not fit the `YYYY` format).

Note: the date format is intentionally different from the format on the HP-48.

## Date display format

By default a Date is displayed on the stack like `Sun 30/Nov/2025`.

> The format can be adjusted by changing the [Time, Date related flags](## Time, Date related flags) or

Reviewer: Wrong markdown link, should be `(#time-date-related-flags)`

Author: already fixed

[Time, Date related settings](## Time, Date related settings)

## Date+Time entry / edit format

A real number with a `_date` unit attached is interpreted as a Date + Time
according to the format `YYYYMMDD.HHMMSS` 

## Date+Time display format

By default a Date+Time is displayed on the stack like `Sun 30/Nov/2025, 16:43:10`.

> The format can be adjusted by changing the [Time, Date related flags](## Time, Date related flags) or / and

Reviewer: In Markdown, you cannot refer to an internal link that way. That would be `(#time-date-related-flags)`.

Author: Fixed that before (after learning that lesson already).

## Date display format

By default a Date is displayed on the stack like `Sun 30/Nov/2025`.

> The format can be adjusted by changing the [Time, Date related flags](#time,-date-related-flags) or

Reviewer: Please squash the various tryout commits

> To compute the Julian Day Number for the first day of the millenium:

Reviewer: Comment about SOFA integration, asked Grok about it, no apparent concern regarding license compatibility…

Integrating the IAU SOFA (Standards of Fundamental Astronomy) library with a GPLv3-licensed project is possible, as SOFA's license is permissive and compatible with GPLv3. SOFA provides a collection of routines for fundamental astronomy calculations, such as time scales, Earth orientation, astrometry, and calendars. It's available in ANSI C and Fortran 77 implementations. Below, I'll outline the integration process and address licensing concerns.

### Integration Steps
SOFA is distributed as source code, so you'll typically compile it into your project or build it as a static/shared library. Here's a high-level guide:

1. **Download SOFA**:
   - Visit the official website (https://www.iausofa.org) and navigate to the "Current Software" section.
   - Select the version for your language: ANSI C (e.g., sofa_c-20231011.tar.gz) or Fortran 77 (e.g., sofa_f-20231011.tar.gz). The latest release as of your query date is from October 2023, but check for updates.
   - Extract the archive. It includes source files (e.g., *.c and *.h for C), documentation (e.g., sofa.pdf), and test programs.

2. **Build SOFA**:
   - SOFA doesn't use a build system like CMake; it's a set of standalone routines.
   - For C: Compile the source files (e.g., using GCC: `gcc -c *.c`) to create object files (*.o), then archive them into a static library (e.g., `ar rcs libsofa.a *.o`).
   - For Fortran: Similar process with a Fortran compiler (e.g., gfortran: `gfortran -c *.f`, then `ar rcs libsofa.a *.o`).
   - Optionally, build a shared library (e.g., `gcc -shared -o libsofa.so *.o`).
   - No external dependencies are required beyond a standard compiler.

3. **Incorporate SOFA into Your Project**:
   - If your GPLv3 project is in C/C++: Include SOFA headers (e.g., `#include "sofa.h"`) and link against the library (e.g., in a Makefile: `gcc your_program.c -L. -lsofa -lm`).
   - For Fortran projects: Include the routines directly or link similarly.
   - For other languages (e.g., Python via ctypes, or wrappers like pysofa): Use existing bindings if available, or create your own. Ensure any wrappers comply with SOFA's terms.
   - Call routines as needed. For example, in C: Use functions like `iauTaiutc()` for time conversions or `iauAtci13()` for astrometry. Refer to the manual (sofa.pdf) for API details, parameters, and examples.
   - Test: Compile and run the provided validation programs (e.g., t_sofa_c.c) to verify your build.

For detailed examples, consult the SOFA manual included in the download, which covers usage, precision notes, and validation. Community resources like GitHub forks (e.g., liberfa/erfa for a renamed, modifiable version) or astronomy software projects (e.g., Astropy) demonstrate integrations.

### Licensing Concerns
- **SOFA License Overview**: SOFA is free for any use (including commercial) under its custom license, owned by the IAU SOFA Board. It allows copying, distribution, and adaptation without royalties, but with conditions to preserve its status as an IAU standard.
- **Compatibility with GPLv3**: Yes, it's compatible. You can include unmodified SOFA code in your GPLv3 project and distribute the combined work under GPLv3. The SOFA license doesn't impose conflicting restrictions (e.g., no anti-copyleft clauses). However:
  - If you modify SOFA code, it becomes a "derived work." You must: (i) mark it as derived and not endorsed by SOFA; (ii) describe changes in the source; (iii) rename routines (remove "iau" or "sofa" prefixes); (iv) not misrepresent origins or claim patents on SOFA parts; (v) reproduce these requirements in your distribution. Modifications are discouraged to maintain standard compliance.
  - For unmodified use (recommended), simply include the original source or binaries, retaining copyright notices.
  - Acknowledgment: In publications or products, cite SOFA (e.g., "Software Routines from the IAU SOFA Collection: Copyright © International Astronomical Union Standards of Fundamental Astronomy (http://www.iausofa.org)").
- **Potential Issues**:
  - No warranties: SOFA is "as is," so your project inherits this for SOFA parts.
  - If distributing binaries, provide SOFA source as required by GPLv3, and comply with SOFA's terms (e.g., no disrepute via misuse).
  - For pure compatibility, consider ERFA (Essential Routines for Fundamental Astronomy), a SOFA derivative with renamed functions under a BSD-like license, which is easier for heavy modifications.
- **Recommendations**: Review the full license (available on the SOFA website). If your project involves significant changes, use ERFA instead. Consult the FSF's license compatibility guide or a legal expert for complex cases, but unmodified integration is straightforward and low-risk.

If you need help with specific routines or encounter issues, the SOFA email support (sofa@ukho.gov.uk) or astronomy forums are useful.

Author: I take from this that you are Ok with using the SOFA routines.

All the build stuff, like makefiles, need to include the 4 sofa files.
I still don't see me doing that.

```rpl
20000101 JDN

> @ Expecting 2 451 544 ¹/₂

Reviewer: Can't say that I like this at all 😄

Author: Julian days start at noon.
it is what it is.
